module/zfs/dmu.c

```diff
@@ -1929,6 +1930,18 @@ dmu_object_size_from_db(dmu_buf_t *db_fake, uint32_t *blksize,
 }
```

Comment:
Looks like dmu_object_size_from_db() also needs to be updated.  The "add 1 for dnode space" is incorrect for large dnodes, so this should be:

        /* add in the number of slots used for the dnode itself */
        *nbk512 = ... SPA_MINBLOCKSHIFT) + (dn->dn_num_slots << DNODE_SHIFT);